File: sdks/go/cmd/elydora/plugins/kimi_io.go
package plugins

import (
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"syscall"
)

func kimiHomeDirectory() (string, error) {
	home, err := os.UserHomeDir()
	if err != nil {
		return "", fmt.Errorf("resolve home directory: %w", err)
	}
	return home, nil
}

func isMissingKimiPath(err error) bool {
	return os.IsNotExist(err) || errors.Is(err, syscall.ENOTDIR)
}

func kimiEntryExists(path, label string) (bool, error) {
	_, err := os.Lstat(path)
	if err == nil {
		return true, nil
	}
	if isMissingKimiPath(err) {
		return false, nil
	}
	return false, fmt.Errorf("inspect %s at %s: %w", label, path, err)
}

func stableKimiContract(configPath string) kimiContract {
	return kimiContract{
		generation: "stable", runtimeName: "Kimi Code",
		label: "Kimi Code hooks config", directoryLabel: "Kimi Code home directory",
		configPath: configPath, events: kimiModernEvents,
	}
}

func legacyKimiContract(configPath string) kimiContract {
	return kimiContract{
		generation: "legacy", runtimeName: "kimi-cli",
		label:          "kimi-cli legacy hooks config",
		directoryLabel: "kimi-cli legacy home directory",
		configPath:     configPath, events: kimiSharedEvents,
	}
}

func resolveKimiContracts() ([]kimiContract, error) {
	home, err := kimiHomeDirectory()
	if err != nil {
		return nil, err
	}
	configuredHome := os.Getenv("KIMI_CODE_HOME")
	stableHome := filepath.Join(home, ".kimi-code")
	explicitHome := configuredHome != ""
	if explicitHome {
		stableHome, err = filepath.Abs(configuredHome)
		if err != nil {
			return nil, fmt.Errorf("resolve KIMI_CODE_HOME at %s: %w", configuredHome, err)
		}
	}
	legacyHome := filepath.Join(home, ".kimi")
	stable := stableKimiContract(filepath.Join(stableHome, "config.toml"))
	legacy := legacyKimiContract(filepath.Join(legacyHome, "config.toml"))
	if sameKimiPath(stable.configPath, legacy.configPath) {
		return []kimiContract{stable}, nil
	}

	stableDetected := explicitHome
	if !stableDetected {
		stableDetected, err = kimiEntryExists(stableHome, "Kimi Code home")
		if err != nil {
			return nil, err
		}
	}
	legacyDetected, err := kimiEntryExists(legacyHome, "kimi-cli legacy home")
	if err != nil {
		return nil, err
	}
	if legacyDetected && !stableDetected {
		return []kimiContract{legacy}, nil
	}
	if legacyDetected {
		return []kimiContract{stable, legacy}, nil
	}
	return []kimiContract{stable}, nil
}

func readKimiConfig(contract kimiContract) (kimiDocument, error) {
	directory := filepath.Dir(contract.configPath)
	if _, err := managedPhysicalDirectoryExists(directory, contract.directoryLabel); err != nil {
		return kimiDocument{}, err
	}
	snapshot, err := readManagedFile(
		contract.configPath,
		contract.label,
		maxManagedSourceBytes,
	)
	if err != nil {
		return kimiDocument{}, err
	}
	if snapshot == nil {
		return parseKimiDocument(contract, nil, false)
	}
	return parseKimiDocument(contract, snapshot.contents, true)
}

func readAllKimiConfigs() ([]kimiDocument, error) {
	contracts, err := resolveKimiContracts()
	if err != nil {
		return nil, err
	}
	documents := make([]kimiDocument, 0, len(contracts))
	for _, contract := range contracts {
		document, err := readKimiConfig(contract)
		if err != nil {
			return nil, err
		}
		documents = append(documents, document)
	}
	return documents, nil
}

func kimiRuntimeNames(documents []kimiDocument) string {
	names := make([]string, 0, len(documents))
	for _, document := range documents {
		names = append(names, document.contract.runtimeName)
	}
	return strings.Join(names, " and ")
}

File: sdks/node/src/plugins/kimi-io.ts
import fsp from 'node:fs/promises';
import os from 'node:os';
import path from 'node:path';
import { asError, errorMessage, hasCode, samePath } from './common.js';
import {
  AGENT_KEY,
  LEGACY_EVENTS,
  STABLE_EVENTS,
  createKimiDocument,
  parseKimiDocument,
  type KimiConfigDocument,
  type KimiContract,
  type KimiRuntimeContract,
} from './kimi-contract.js';
import { inspectPhysicalDirectory, readPhysicalFile } from './managed-files.js';
import { managedRuntimeFilesExist } from './managed-runtime-status.js';

async function pathEntryExists(filePath: string, label: string): Promise<boolean> {
  try {
    await fsp.lstat(filePath);
    return true;
  } catch (error) {
    if (hasCode(error, 'ENOENT') || hasCode(error, 'ENOTDIR')) return false;
    throw new Error(`Inspect ${label} at ${filePath}: ${errorMessage(error)}`, {
      cause: asError(error),
    });
  }
}

function stableContract(configPath: string): KimiContract {
  return {
    generation: 'stable',
    runtimeName: 'Kimi Code',
    label: 'Kimi Code hooks config',
    directoryLabel: 'Kimi Code home directory',
    configPath,
    events: STABLE_EVENTS,
  };
}

function legacyContract(configPath: string): KimiContract {
  return {
    generation: 'legacy',
    runtimeName: 'kimi-cli',
    label: 'kimi-cli legacy hooks config',
    directoryLabel: 'kimi-cli legacy home directory',
    configPath,
    events: LEGACY_EVENTS,
  };
}

export async function resolveKimiContracts(): Promise<KimiContract[]> {
  const home = os.homedir();
  const configuredHome = process.env.KIMI_CODE_HOME;
  const explicitHome = configuredHome === undefined || configuredHome === ''
    ? undefined
    : path.resolve(configuredHome);
  const stableHome = explicitHome ?? path.join(home, '.kimi-code');
  const legacyHome = path.join(home, '.kimi');
  const stable = stableContract(path.join(stableHome, 'config.toml'));
  const legacy = legacyContract(path.join(legacyHome, 'config.toml'));
  if (samePath(stable.configPath, legacy.configPath)) return [stable];

  const stableDetected = explicitHome !== undefined
    || await pathEntryExists(stableHome, 'Kimi Code home');
  const legacyDetected = await pathEntryExists(legacyHome, 'kimi-cli legacy home');
  if (legacyDetected && !stableDetected) return [legacy];
  return legacyDetected ? [stable, legacy] : [stable];
}

async function readKimiDocument(contract: KimiContract): Promise<KimiConfigDocument> {
  await inspectPhysicalDirectory(path.dirname(contract.configPath), contract.directoryLabel);
  const snapshot = await readPhysicalFile(contract.configPath, contract.label);
  return snapshot
    ? parseKimiDocument(contract, snapshot.contents)
    : createKimiDocument(contract);
}

export async function readKimiDocuments(): Promise<KimiConfigDocument[]> {
  const contracts = await resolveKimiContracts();
  const documents: KimiConfigDocument[] = [];
  for (const contract of contracts) documents.push(await readKimiDocument(contract));
  return documents;
}

export async function kimiRuntimeFilesExist(
  contracts: readonly KimiRuntimeContract[],
): Promise<boolean> {
  for (const contract of contracts) {
    if (await managedRuntimeFilesExist(contract, AGENT_KEY)) return true;
  }
  return false;
}
